lib/utils/docker.coffee

```diff
@@ -0,0 +1,201 @@
+		# good, local docker socket
+		connectOpts['socketPath'] = opts.docker
+	else if opts.dockerHost? and not opts.docker?
+		# Good a host is provided, and local socket isn't
+		connectOpts['host'] = opts.dockerHost
+		connectOpts['port'] = opts.dockerPort || 2376
+	else if opts.docker? and opts.dockerHost?
+		# Both provided, no obvious way to continue
+		throw new Error("Both a local docker socket and docker host have been provided. Don't know how to continue.")
+	else
+		# None provided, assume default docker local socket
+		connectOpts['socketPath'] = '/var/run/docker.sock'
+
+	# Now need to check if the user wants to connect over TLS
+	# to the host
+
+	# If any are set...
+	if (opts.ca? or opts.cert? or opts.key?)
+		# but not all
+		if not (opts.ca? and opts.cert? and opts.key?)
+			throw new Error('You must provide a CA, certificate and key in order to use TLS')
+		connectOpts['ca'] = opts.ca
+		connectOpts['cert'] = opts.cert
+		connectOpts['key'] = opts.key
+
+	return connectOpts
+
+exports.tarDirectory = tarDirectory = (dir) ->
+	Promise = require('bluebird')
+	tar = require('tar-stream')
+	klaw = require('klaw')
+	path = require('path')
+	fs = require('mz/fs')
+	streamToPromise = require('stream-to-promise')
+
+	getFiles = ->
+		Promise.resolve(streamToPromise(klaw(dir)))
+			.filter((item) -> not item.stats.isDirectory())
+			.map((item) -> item.path)
+
+	pack = tar.pack()
+	getFiles(dir)
+	.map (file) ->
+		relPath = path.relative(path.resolve(dir), file)
+		Promise.all([relPath, fs.stat(file), fs.readFile(file)])
+	.map ([filename, stats, data]) ->
+		pack.entryAsync({ name: filename, size: stats.size }, data)
+	.then ->
+		pack.finalize()
+		return pack
+
+# Pass in the command line parameters and options and also
+# a function which will return the information about the bundle
+exports.runBuild = (params, options, getBundleInfo) ->
+	Promise = require('bluebird')
+	dockerBuild = require('resin-docker-build')
+	resolver = require('resin-bundle-resolve')
+
+	# The default build context is the current directory
+	params.context ?= '.'
+
+	# Tar up the directory, ready for the build stream
+	tarDirectory(params.context)
+	.then (tarStream) ->
+		new Promise (resolve, reject) ->
+			hooks =
+				buildSuccess: (image, layers) ->
+					if options.tag?
+						console.log("Tagging image as #{options.tag}")
+					resolve(image)
+				buildFailure: reject
+				buildStream: (stream) ->
+					info = getBundleInfo(options)
+					# info is either a promise, or is undefined
+					if info?
+						info.then ([arch, deviceType]) ->
+							# Perform type resolution on the project
+							bundle = new resolver.Bundle(tarStream, deviceType, arch)
+							resolver.resolveBundle(bundle, resolver.getDefaultResolvers())
+						.then (resolved) ->
+							console.log("Building #{resolved.projectType} project")
+							# Send the resolved tar stream to the docker daemon
+							resolved.tarStream.pipe(stream)
+						.catch(reject)
```

Comment:
It seems to work, if the resolution fails, an error throws and the promise rejects, and it hits the global handler happily. Which part are you not sure on?